src/transformers/models/wav2vec2/modeling_wav2vec2.py

```diff
@@ -67,50 +67,43 @@ def _compute_mask_indices(
     Adapted from `fairseq's data_utils.py
     <https://github.com/pytorch/fairseq/blob/e0788f7007a8473a76db573985031f3c94201e79/fairseq/data/data_utils.py#L376>`__.
     """
+    torch_device = "cuda" if torch.cuda.is_available() else "cpu"
     bsz, all_sz = shape
-    mask = np.full((bsz, all_sz), False)
-
-    all_num_mask = int(
-        # add a random number for probabilistic rounding
-        mask_prob * all_sz / float(mask_length)
-        + np.random.rand()
-    )
-
-    all_num_mask = max(min_masks, all_num_mask)
-
-    mask_idcs = []
+    mask = torch.full((bsz, all_sz), False, device=torch_device)
+    min_len_unique, mask_idcs = float("Inf"), []
     padding_mask = attention_mask.ne(1) if attention_mask is not None else None
+    if padding_mask is not None:
+        sz = all_sz - padding_mask.long().sum(dim=1)
+        num_mask = (((sz * mask_prob) / mask_length) + torch.rand(bsz, device=torch_device)).int()
+    else:
+        sz = torch.full([bsz], all_sz, device=torch_device)
+        all_num_mask = ((all_sz * mask_prob) / mask_length) + torch.rand(1, device=torch_device)[0]
+        num_mask = torch.full([bsz], all_num_mask, device=torch_device).int()
+
+    num_mask = torch.clamp(num_mask, min=min_masks)
     for i in range(bsz):
-        if padding_mask is not None:
-            sz = all_sz - padding_mask[i].long().sum().item()
```

Comment:
I think we need to add this code in the loop as well -> `num_mask` depends on the padding_mask which in the new code doesn't seem to be the case

Reply:
In the new code as well as old code the padding mask was calculated beforehand outside the loop.
As padding mask is calculated before the for loop and isn't changing inside the loop we can vectorize that calculation.

In for loop implementation, we calculated the sz and num_mask for each row inside loop.
In the vectorized implementation, we can calculate values of sz and num_mask before for loop.